Feature/8 apply mybatis to user

jieun-dev1 opened the pull request:
[0921] 
+수정사항 반영

-type-handlers 를 없애고, underscore를 camelcase로 직접 매핑이 아니라 resultMap에서 처리함.
-기본 생성자를 추가하여 마이바티스에서 객체 바인딩을 제대로 실행할 수 있도록 함.
-findEmailDuplicate 중복이어서 삭제 
-search와 enumCheck는 불필요하다고 판단하여 삭제 

아래 테스트 내역 모두 동작함을 확인. 

[0919] 

[개요] 
My Batis 적용하였습니다. 

[수정] 
-Mybatis에서는 Mapper가 인터페이스 이기 때문에, 구현을 위해 반환 타입을 없애고  Exception 의 위치(Repo->Service)를 수정. 
-LoginUser객체가 Password가 필요치 않기 때문에 ( 로그인 될 때만 필요한 정보). 대신 Role을 추가함.기능에 따라 권한 관리가 필요하기 때문)

[테스트 내역] 

*MyBatisRepository & InMemoryRepository로 작업 시 동작 여부 확인

-회원 가입이 정상적으로 되는 것 확인 
-중복 아이디는 400Exception 확인 
-로그인 정상적으로 되는 것 확인 (비밀 번호 일치시에만)
-로그인 시에만 home이 접근 가능하고, 안되면 Unauthorized 발생을 확인.

jieun-dev1 committed:
chore: jar파일이 gitignore에 등록되어 gradle wrapper 파일이 누락됨. git ignore 수정하고 push 재반영.

jieun-dev1 committed:
Feat:MyBatis도입

-Exception이 Service 로직에서 발생하도록변경(MyBatis적용시,Mapper 인터페이스에서 구현 할 수 없기 때문)
-UserRepository는 저장소에 저장되므로,User 객체 반환이 필요치 않아서 void return type 으로바꿈.
-Mapper는 인터페이스이므로,어댑터 패턴을 사용하여 객체를 변환하고 UserRepository를 구현하게함

jieun-dev1 committed:
Feat: MyBatis enum 처리를 위해 enum 타입 핸들러 구현
-enum type 상수가 아닌 값이 파라미터에 들어올 경우, 현재 500에러. 별도 예외 처리 진행 중.

jieun-dev1 committed:
Feat: 주석 추가, 간격 등 수정
checkstyle warning 으로 변경

jieun-dev1 committed:
Feat: LoginUser 객체가 Password 가 필요치 않기 때문에 (한번 로그인 되면 Password 는 다른 기능에서 요구하지 않는 정보임. 대신 Role을 추가함. 기능에 따라 권한 관리가 필요하기 때문)

f-lab-bright commented on src/main/java/com/example/freshcart/domain/Role.java:
```diff
@@ -4,5 +4,26 @@
-    USER, SELLER, ADMIN
+  USER, SELLER, ADMIN;
+
+  public static Role search(String columnName) {
+    for (Role element : Role.values()) {
+      if (element.name().equalsIgnoreCase(columnName)) {
+        return element;
+      }
+    }
+    return null;
+  }
+
+  public static Role enumCheck(String name) {
+    Role result = null;
+    for (Role role : values()) {
+      //일치한다면 반복문을 빠져나올 수 있다.
+      if (role.name().equalsIgnoreCase(name)) {
+        result = role;
+        break;
+      }
+    }
+    return result;
+  }
```
두 함수의 역할이 중복되지 않나요??
그리고 `Role.valueOf("ADMIN")` 으로 사용할 수 있는 이미 만들어져있는 함수가 있을거에요

jieun-dev1 replied:
네! 이 코드는 작업 중이던건데 실수로 푸시했네요ㅜ 삭제 반영했습니다. admin 권한 확인은 제품 등록 기능 만들 때 반영할게요.

f-lab-bright commented on src/main/java/com/example/freshcart/domain/UserRepository.java:
```diff
@@ -7,9 +7,9 @@
+  void save(User user);
 
-  void findEmailDuplicate(String email);
+  User findEmailDuplicate(String email);
```
이러면 아래에 정의한 함수와 다를바가 없어지네요.
`User findByEmail(String email)` 하나만 있으면 될 것 같네요.

f-lab-bright commented on src/main/java/com/example/freshcart/domain/UserRepository.java:
```diff
@@ -7,9 +7,9 @@
  */
 public interface UserRepository {
-  User save(User user);
+  void save(User user);
```
save 시에 반환하지 않아도 괜찮을까요?

jieun-dev1 replied:
아직 반환 객체(apiReponse)를 만들지 않았지만, 모든 요청에 대해 일관된 응답을 해야합니다. 보통 가입 후에 OO 님 환영합니다 같은 응답 메시지가 나오는 경우가 있는데요. 응답 객체를 만들 때, 유저 인식 정도 (email이나 이름 등) 을 반환하는 게 좋을 것 같네요!

f-lab-bright commented on src/main/java/com/example/freshcart/infrastructure/UserMapper.java:
```diff
@@ -0,0 +1,18 @@
+@Mapper
+public interface UserMapper {
+
+  void save(User user);
+
+  User findEmailDuplicate(String email);
+
+  Optional<User> findByUserEmail(String email);
+
+  List<User> findAll();
```
여기는 repository가 아니라 mybatis의 영역이고, mybatis는 query mapper에 가까우니, database의 동작스러운 메서드명이 더 이해하기 좋을 수 있습니다.
find -> select
save -> insert | update (이 둘이 구분 되어야 함)

f-lab-bright commented on src/main/java/com/example/freshcart/infrastructure/config/RepositoryConfig.java:
```diff
@@ -12,8 +14,15 @@
 public class RepositoryConfig {
 
+  private final UserMapper userMapper;
+
+  public RepositoryConfig(UserMapper userMapper) {
+    this.userMapper = userMapper;
+  }
+
   @Bean
   public UserRepository userRepository() {
-    return new UserInMemoryRepository();
+    return new UserMapperRepositoryAdaptor(userMapper);
```
여러번 쓰이는게 아니라면 userMapper를 필드로 주입받아서 사용할 필요도 없고, 
`UserRepository userRepository(UserMapper userMapper) { ... }` 로 사용하시면 됩니다.

f-lab-bright commented on src/main/java/com/example/freshcart/infrastructure/config/UserMapperRepositoryAdaptor.java:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  @Override
+  public void save(User user) {
+    userMapper.save(user);
+  }
```
여기서 User가 처음 만들어지는 경우 id (user table의 pk) 가 auto_increment로 만들어질텐데, id를 알 수가 있나요??
User가 만들어졌을 때 만들어진 User의 id를 알 수 있어야 합니다.
모든 api에서 무언가를 생성했을 때는 생성한 녀석의 id를 항상 알 수 있어야 다음 액션을 할 수 있는 상황들이 자주 있습니다.

f-lab-bright commented on src/main/java/com/example/freshcart/infrastructure/handlers/RoleTypeHandler.java:
```diff
@@ -0,0 +1,33 @@
+import org.apache.ibatis.type.BaseTypeHandler;
+import org.apache.ibatis.type.JdbcType;
+
+public class RoleTypeHandler extends BaseTypeHandler<Role> {
```
이렇게 사용하면 enum을 사용할 때 마다 매번 새로운 TypeHandler를 만들어서 사용해야 하는데, 적절한 방법일까요??

jieun-dev1 replied:
다시 공부해보니, MyBatis는 기본값으로 EnumTypeHandler를 사용하고, 이는 enum 값을 상수로 바꿔줍니다. 
이 경우는 상수만 있는 단일 Enum 이기 때문에, TypeHandler를 만들어줄 필요가 없네요. 
By default, MyBatis uses EnumTypeHandler to convert the Enum values to their names  (출처: https://mybatis.org/mybatis-3/configuration.html)

f-lab-bright commented on src/main/resources/mapper/user-mapper.xml:
```diff
@@ -0,0 +1,26 @@
+<!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN"
+  "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
+<mapper namespace="com.example.freshcart.infrastructure.UserMapper">
+  <insert id = "save" keyProperty="id" useGeneratedKeys="true">
```
keyProperty와 useGeneratedKeys가 어떤 의미일까요?

jieun-dev1 replied:
useGeneratedKeys는 MyBatis에게 JDBC getGeneratedKeysMethod를 쓰라고 알려줍니다. 이 메서드를 쓰면, SQL 내부에서 생성된 id(ex Auto_increment)가 반환할 수 있습니다. 

이때, 어떤 ID 인지 식별해야 할텐데, keyProperty에  ID 에 해당하는 객체 properpty의 이름을 지정하면 됩니다. 
그러면 MyBatis가 db 에서 가져온 key value를 (getGeneratedKeysMethod를 통해) 객체 property에 set 해줄 수 있습니다. 

https://mybatis.org/mybatis-3/sqlmap-xml.html

f-lab-bright commented on src/main/resources/mapper/user-mapper.xml:
```diff
@@ -0,0 +1,26 @@
+<!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN"
+  "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
+<mapper namespace="com.example.freshcart.infrastructure.UserMapper">
+  <insert id = "save" keyProperty="id" useGeneratedKeys="true">
```
sql으로 insert와 update를 둘 다 처리 해야할텐데, save라는 이름이 적절할까요?

jieun-dev1 committed:
chore: 필요없는 파일 삭제
MyBatis는 기본값으로 EnumTypeHandler를 사용하고, 이는 enum 값을 상수로 바꿔줌.
이 경우는 상수만 있는 단일 Enum 이기 때문에, TypeHandler를 만들어줄 필요가 없어서 삭제.

jieun-dev1 committed:
chore: 패키지 이동

jieun-dev1 committed:
feature: findEmailDuplicate과 findByUserEmail로직이 중복되어, findByUserEmail만 남김

-기존 Optional에서 User 객체로 바꿔서, 서비스에서 null 처리를 하도록 함.

jieun-dev1 committed:
feature: enum 처리
-type-handlers 를 없애고, underscore를 camelcase로 직접 매핑이 아니라 resultMap에서 처리함.
-기본 생성자를 추가하여 마이바티스에서 객체 바인딩을 제대로 실행할 수 있도록 함.
-search와 enumCheck는 불필요하다고 판단하여 삭제

jieun-dev1 committed:
feature: UserMapper는 UserMapperRepositoryAdaptor에서 final 주입해서 사용

f-lab-bright commented on src/main/resources/mapper/user-mapper.xml:
```diff
@@ -0,0 +1,27 @@
+    values (#{email}, #{password}, #{phoneNumber}, #{name}, #{role});
+  </insert>
+
+  <resultMap id="userResultMap" type="User">
```
가독성을 위해 resultMap을 최상단에 두는게 더 좋겠네요 :)

f-lab-bright commented on src/main/java/com/example/freshcart/infrastructure/UserMapperRepositoryAdaptor.java:
```diff
@@ -0,0 +1,39 @@
+
+  @Override
+  public User save(User user) {
+    userMapper.insert(user);
```
수정이 일어나는 경우에도 이 메소드가 호출이 되면 동작하지 않겠네요.
어떻게 하는게 더 좋을지 고민해보시면 좋겠네요 :)

jieun-dev1 replied:
회원 정보 변경 말씀이신가요? 그렇다면 회원 가입과 회원 정보 수정 API 는 별도로 둘 생각이었는데요 (현재 기능 개발의 우선순위는 아니지만) 가입 api 에서 수정도 고려해야 하는 이유가 있을까요?

f-lab-bright replied:
api는 분리 되지만, repository의 save라는 메서드는 두 곳에서 다 사용될테니까요 :)

f-lab-bright reviewed (approved):
LGTM with Comments

jieun-dev1 committed:
chore: 가독성을 위해 resultMap을 최상단에 둠
